RFC-0004-functions-stats.md

```diff
@@ -0,0 +1,472 @@
+
+  /// map of function position argument to source statistics transformation.
+  std::map<int, VectorFunctionPropagateSourceStats> transformSourceStats;
+};
+```
+
+Where `VectorFunctionConstantStatistics` and `VectorFunctionPropagateSourceStats` is  defined as follows :
+
+```c++
+struct VectorFunctionPropagateSourceStats {
+
+  bool propagateAllStats{true};
+  
+  StatsPropagationBehavior lowValue{UNKNOWN};
+
+  StatsPropagationBehavior highValue{UNKNOWN};
+
+  StatsPropagationBehavior nullFraction{UNKNOWN};
+
+  StatsPropagationBehavior averageRowSize{UNKNOWN};
+
+  StatsPropagationBehavior distinctValuesCount{UNKNOWN};
+};
+
+struct VectorFunctionConstantStatistics {
+  double lowValue{NAN};
+
+  double highValue{NAN};
+
+  double nullFraction{NAN};
+
+  double averageRowSize{NAN};
+
+  double distinctValuesCount{NAN};
+};
+```
+
+`StatsPropagationBehavior` is an enum that defines different ways stats for functions can be estimated.
+
+```c++
+namespace facebook::velox::exec {
+enum StatsPropagationBehavior {
+  /** Use the max value across all arguments to derive the new stats value */
+  USE_MAX_ARGUMENT,
+  /** Sum the stats value of all arguments to derive the new stats value */
+  SUM_ARGUMENTS,
+  /** Propagate the source stats as-is */
+  USE_SOURCE_STATS,
+  /** Use the value of output row count. */
+  ROW_COUNT,
+  /** Use the value of row_count * (1 - null_fraction). */
+  ROW_COUNT_TIMES_INV_NULL_FRACTION,
+  /** use the value of TYPE_WIDTH in varchar(TYPE_WIDTH) */
+  USE_TYPE_WIDTH_VARCHAR,
+  /** Take max of type width of arguments with varchar type. */
+  MAX_TYPE_WIDTH_VARCHAR,
+  /** Stats are unknown and thus no action is performed. */
+  UNKNOWN
+};
+}
+```
+
+Examples for C++ functions:
+
+Following is how metadata would look like for
+[ConcatFunction::metadata](https://github.com/facebookincubator/velox/blob/0adc62e3aaf647596ab476f6300a4165e60f6b91/velox/functions/prestosql/StringFunctions.cpp#L273):
+```c++
+static exec::VectorFunctionMetadata metadata() {
+return {
+    .supportsFlattening = true,
+    .transformSourceStats = {
+        {0,
+         exec::VectorFunctionPropagateSourceStats{
+             .propagateAllStats= false,
+             .nullFraction = exec::SUM_ARGUMENTS,
+             .averageRowSize = exec::SUM_ARGUMENTS}}}};
+}
+```
+
+#### How does a C++ worker communicate functions stats behavior and constant stats to the coordinator.
+
+The ongoing work described in [RFC-0003](RFC-0003-native-spi.md), with function registries already includes function metadata as field .
+Since we are introducing fields inside the
+[FunctionMetadata.h](https://github.com/facebookincubator/velox/blob/0adc62e3aaf647596ab476f6300a4165e60f6b91/velox/expression/FunctionMetadata.h)
+the new fields have to be sent as is for further processing to the coordinator.
+
+In both Java and C++, builtin functions have extra metadata defined in FunctionMetadata class, which is used while computing estimated stats
```

Comment:
The side-car returns https://github.com/prestodb/presto/blob/master/presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/json/JsonBasedUdfFunctionMetadata.java for all Velox functions to the co-ordinator. We would need to enhance this as well for the new Stats. Please add to this RFC.